transactor/src/main/scala/io/mediachain/transactor/JsonSerialization.scala

```diff
@@ -0,0 +1,177 @@
+package io.mediachain.transactor
+
+import io.atomix.catalyst.buffer.{BufferInput, BufferOutput}
+import io.atomix.catalyst.serializer.{Serializer, TypeSerializer}
+import io.mediachain.transactor.Dummies.DummyReference
+import io.mediachain.transactor.Types._
+import org.json4s._
+import org.json4s.jackson.JsonMethods.{compact, render, parse}
+
+/**
+  * Represents a 'multihash' link to an IPFS resource
+  *
+  * @param multihash - the base58 encoded string representation of an IPFS
+  *             multihash
+  */
+case class IPFSReference(multihash: String) extends Reference
+
+object JsonSerialization {
+
+  /** Like json4s ShortTypeHints, but removes the outer class name for
+    * inner classes.  So instead of e.g. `Types$Artefact`, produces
+    * `Artefact`
+    *
+    * These hints are encoded in the `type` field of the serialized json
+    * representation.
+    */
+  private case class ShortTypeHintsIgnoreOuterClass(hints: List[Class[_]])
+    extends TypeHints {
+    def hintFor(clazz: Class[_]) =
+      clazz.getName.substring(clazz.getName.lastIndexOf(".")+1)
+        .split('$').toList.last
+
+    def classFor(hint: String) = hints find (hintFor(_) == hint)
+  }
+
+
+  val copycatSerializers: List[CopycatSerializer[_]] = List(
+    // reference
+    new CopycatSerializer[IPFSReference],
+    new CopycatSerializer[DummyReference],
+
+    // chain references
+    new CopycatSerializer[EntityChainReference],
+    new CopycatSerializer[ArtefactChainReference],
+
+    // canonical records
+    new CopycatSerializer[Entity],
+    new CopycatSerializer[Artefact],
+
+    // chain cells
+    new CopycatSerializer[EntityChainCell],
+    new CopycatSerializer[ArtefactChainCell],
+
+    // journal entries
+    new CopycatSerializer[CanonicalEntry],
+    new CopycatSerializer[ChainEntry],
+
+    // error types
+    new CopycatSerializer[JournalCommitError]
+  )
+
+  private val shortTypeHints =
+    ShortTypeHintsIgnoreOuterClass(copycatSerializers.map(_.serializableClass))
+
+  /**
+    * Rename the `multihash` field to `@link`, to match the IPLD spec
+    */
+  private val referenceSerializer = FieldSerializer[IPFSReference](
+    FieldSerializer.renameTo("multihash", "@link"),
+    FieldSerializer.renameFrom("@link", "multihash")
+  )
+
+  /**
+    * Tells json4s to use the type hints we specified, and to store them
+    * in the `type` field.  Also adds the field-renaming serializer for
+    * `IPFSReference`
+    */
+  implicit val jsonFormats = new DefaultFormats {
+    override val dateFormat = DefaultFormats.lossless.dateFormat
+    override val typeHints = shortTypeHints
+    override val typeHintFieldName: String = "type"
+  } + referenceSerializer
+
+
+  // Conversion functions from transactor types to json4s `JObject`s.
+
+  def toJObject(record: Record): JObject =
+    Extraction.decompose(record).asInstanceOf[JObject]
+
+  def toJObject(chainCell: ChainCell): JObject =
```

Comment:
here there is the intricacy of the chain pointer representation -- None must still generate an {"@iink"=...} pointer to the Nil object.